### PR TITLE
crypto: add go-based BatchVerifier implementation

### PR DESCRIPTION
**Pull request by cce**
## Summary

This adds a pure-Go ed25519 BatchVerifier implementation using on the [ed25519consensus](https://github.com/hdevalence/ed25519consensus) library, plus additional checks to preserve our ed25519 validation criteria, last updated in #3031 (reject small order A; reject non-canonical A, R, S; use the cofactor equation).

Like our [libsodium batch verification implementation](https://github.com/algorand/go-algorand/blob/7804433722d4e919a01d7c585c33eeeb9f716351/crypto/libsodium-fork/src/libsodium/crypto_core/ed25519/ref10/ed25519_ref10.c#L1165-L1204), the `IsCanonicalY` check here is also based on the "Taming the Many EdDSAs" paper in https://eprint.iacr.org/2020/1244
image
image
image

## Test Plan

New tests added to compare ed25519 criteria results match our existing libsodium- and ed25519-donna-based batch verification implementation (from #3031 and defined in https://github.com/algorandfoundation/specs/pull/60). New test helpers run test vectors with different batch sizes, taken from
- 12 edge cases from ["Taming the many EdDSAs"](https://eprint.iacr.org/2020/1244) Appendix C, Table 6c, also used in our libsodium fork's tests in #3031 
- 1025 successful cases from our libsodium fork's tests, also added in #3031.
- 196 edge cases used to draw the 14x14 grid visualizations from the blog post ["It's 255:19AM. Do you know what your validation criteria are?"](https://hdevalence.ca/blog/2020-10-04-its-25519am/) and used in [ed25519consensus](https://github.com/hdevalence/ed25519consensus/blob/main/zip215_test.go)
- 768 edge cases from the Go crypto package's [crypto/ed25519/ed25519vectors_test.go](https://github.com/golang/go/blob/78ef487a6f936a39e9d4ebf66ac421bb1244a7a9/src/crypto/ed25519/ed25519vectors_test.go#L17-L71)

### CODE REVIEW

File: crypto/gobatchverifier.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+	if p[0] < 237 {
+		return true
+	}
+	for i := 1; i < 31; i++ {
+		if p[i] != 255 {
+			return true
+		}
+	}
+	return (p[31] | 128) != 255
+}
```
Comment (cce) [inline crypto/gobatchverifier.go:127]: For comparison, this is based on https://eprint.iacr.org/2020/1244
image
image
image

**Review by zeldovich — APPROVED** [review]

File: crypto/batchverifier.go [hunk 2]
```diff
@@ -73,14 +73,33 @@ func ed25519_randombytes_unsafe(p unsafe.Pointer, len C.size_t) {
...
+
+// SetEd25519BatchVerifier allows the config package to switch the implementation
+// at startup based on configuration. Pass true to use ed25519consensus, false for libsodium.
+func SetEd25519BatchVerifier(useEd25519Consensus bool) {
+	if useEd25519Consensus {
+		ed25519BatchVerifierFactory = makeEd25519ConsensusBatchVerifier
+	} else {
+		ed25519BatchVerifierFactory = makeLibsodiumBatchVerifier
+	}
+}
```
Comment (jannotti) [inline crypto/batchverifier.go:89]: Does this stay forever, or is it just so you can have tests that compare behavior?

File: crypto/batchverifier.go [hunk 3]
```diff
@@ -73,14 +73,33 @@ func ed25519_randombytes_unsafe(p unsafe.Pointer, len C.size_t) {
...
+// MakeBatchVerifier creates a BatchVerifier instance.
 func MakeBatchVerifier() BatchVerifier {
-	return MakeBatchVerifierWithHint(minBatchVerifierAlloc)
+	return ed25519BatchVerifierFactory(minBatchVerifierAlloc)
 }
 
-// MakeBatchVerifierWithHint creates a cgoBatchVerifier instance. This function pre-allocates
-// amount of free space to enqueue signatures without expanding
+// MakeBatchVerifierWithHint creates a BatchVerifier instance. This function pre-allocates
+// amount of free space to enqueue signatures without expanding.
```
Comment (jannotti) [inline crypto/batchverifier.go]:
```suggestion
// MakeBatchVerifierWithHint creates a BatchVerifier instance. This function pre-allocates
// space to enqueue signatures without expanding.
```

File: crypto/batchverifier.go [hunk 4]
```diff
@@ -73,14 +73,33 @@ func ed25519_randombytes_unsafe(p unsafe.Pointer, len C.size_t) {
 const minBatchVerifierAlloc = 16
 const useSingleVerifierDefault = true
 
-// MakeBatchVerifier creates a BatchVerifier instance with the provided options.
+// ed25519BatchVerifierFactory is the global singleton used for batch signature verification.
+// By default it uses the libsodium implementation. This can be changed during initialization
+// (e.g., by the config package when algod loads) to use the ed25519consensus implementation.
+var ed25519BatchVerifierFactory func(hint int) BatchVerifier = makeLibsodiumBatchVerifier
```
Comment (jannotti) [inline crypto/batchverifier.go:79]: You can call this variable `MakeBatchVerifier` and remove the func by that name.

File: crypto/batchverifier.go [hunk 5]
```diff
@@ -73,14 +73,33 @@ func ed25519_randombytes_unsafe(p unsafe.Pointer, len C.size_t) {
...
-// amount of free space to enqueue signatures without expanding
+// MakeBatchVerifierWithHint creates a BatchVerifier instance. This function pre-allocates
+// amount of free space to enqueue signatures without expanding.
 func MakeBatchVerifierWithHint(hint int) BatchVerifier {
+	return ed25519BatchVerifierFactory(hint)
+}
+
+func makeLibsodiumBatchVerifier(hint int) BatchVerifier {
 	// preallocate enough storage for the expected usage. We will reallocate as needed.
 	if hint < minBatchVerifierAlloc {
```
Comment (jannotti) [inline crypto/batchverifier.go]: I would have expected this to get converted to `hint = max(hint, minBatchVerifierAlloc)` with modernizer.

File: crypto/batchverifier.go [hunk 6]
```diff
@@ -185,18 +204,26 @@ func cgoBatchVerificationImpl(messages []byte, msgLengths []uint64, publicKeys [
...
-		(*C.int)(&valid[0]))
+		(*C.int)(unsafe.SliceData(valid)))
+
+	if allValid == 0 { // all signatures valid
+		return true, nil
+	}
 
+	// not all signatures valid, identify the failed signatures
 	failed = make([]bool, numberOfSignatures)
 	for i := 0; i < numberOfSignatures; i++ {
```
Comment (jannotti) [inline crypto/batchverifier.go:225]: I also though modernizer was changing this to `range numberOfSignatures`

File: crypto/batchverifier_bench_test.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+	secrets := GenerateSignatureSecrets(s)
+	return secrets.SignatureVerifier, msg, secrets.Sign(msg)
+}
+
+// BenchmarkBatchVerifierImpls benchmarks different batch verification implementations
+// with realistic batch sizes (100 batches of 64 signatures each)
+func BenchmarkBatchVerifierImpls(b *testing.B) {
+	partitiontest.PartitionTest(b)
+
+	bN := 100
```
Comment (jannotti) [inline crypto/batchverifier_bench_test.go]: I think you're just prepping some signatures for use in later benchmarks, but calling the variable `bN` makes it seem like you're doing something funky with `b.N`.

File: crypto/batchverifier_bench_test.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+		for i := 0; i < min(b.N, len(msgs)); i++ {
+			for j := range msgs[i] {
+				bv.EnqueueSignature(pks[i][j], msgs[i][j], sigs[i][j])
+			}
+			require.NoError(b, bv.Verify())
+		}
+		b.StopTimer()
+	}
+	b.StopTimer()
+	b.ResetTimer()
```
Comment (jannotti) [inline crypto/batchverifier_bench_test.go]: What do these do when you only care about the sub benchmarks?

File: crypto/batchverifier.go (hunk 5 above)
Comment (jannotti) [reply crypto/batchverifier.go]: I should say though, it seems weird not to trust callers.  If you get a 0, fine, I guess use the global min. But callers who are using a batch to verify a multisig or a consensus vote ought to know what they're doing and call with the right size.

File: crypto/gobatchverifier.go [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+type ed25519ConsensusBatchVerifier struct {
+	entries      []ed25519ConsensusVerifyEntry // used in VerifyWithFeedback to identify failed signatures
+	failedChecks bool                          // true if any entry failed non-canonical or small-order checks
+	bv           ed25519consensus.BatchVerifier
+}
+
+func makeEd25519ConsensusBatchVerifier(hint int) BatchVerifier {
+	if hint < minBatchVerifierAlloc {
+		hint = minBatchVerifierAlloc
+	}
```
Comment (jannotti) [inline crypto/gobatchverifier.go]: Use `max`, or trust caller to be right?

File: crypto/gobatchverifier.go [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+
+	return failed, ErrBatchHasFailedSigs
+}
+
+// Check that Y is canonical, using the succeed-fast algorithm from
+// the "Taming the many EdDSAs" paper.
+func isCanonicalY(p []byte) bool {
+	if len(p) != 32 {
+		return false
+	}
```
Comment (jannotti) [inline crypto/gobatchverifier.go]: Is this here so that Go can optimize away the bounds checks below?

File: crypto/gobatchverifier.go [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+			0xec, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+			0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+			0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+			0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+		},
+	} {
+		if bytes.Equal(p[:], invalidEncoding[:]) {
+			return false
+		}
+	}
```
Comment (jannotti) [inline crypto/gobatchverifier.go]: Is looping instead of making two calls to `bytes.Equal` faster?

File: crypto/gobatchverifier.go (hunk 10 above)
Comment (cce) [reply crypto/gobatchverifier.go]: Originally this function took a `p [32]byte` so I added this when I switched it to []byte ... good point, well if you pass in an empty or <32-length slice you will get a panic as currently written, and if you pass in a >32-length slice you could get a success for an invalid input, but all the code paths that call this already are ensuring the input is 32 bytes, so maybe it's OK to panic. Not sure what is more performant and whether it would optimize away the bounds checks, or whether I could/should switch it back to [32]byte again

File: crypto/batchverifier.go (hunk 2 above)
Comment (cce) [reply crypto/batchverifier.go:89]: Yes I want to make it configurable so I can run tests with it turned on, catch up and run different networks, compare behavior. Then eventually make it the default

File: crypto/batchverifier_bench_test.go (hunk 7 above)
Comment (cce) [reply crypto/batchverifier_bench_test.go]: updated

File: crypto/batchverifier.go (hunk 6 above)
Comment (cce) [reply crypto/batchverifier.go:225]: I never merged that one, it was https://github.com/algorand/go-algorand/pull/6435, it seemed like such a big PR and I was on the fence, but we can bring it back..

File: crypto/batchverifier.go (hunk 5 above)
Comment (cce) [reply crypto/batchverifier.go]: Agreed, in practice running batches of 3 from OneTimeSig is going to be the biggest user of this, updated

File: crypto/batchverifier.go (hunk 4 above)
Comment (cce) [reply crypto/batchverifier.go:79]: That makes it settable by any other package, though ..

File: crypto/gobatchverifier.go (hunk 9 above)
Comment (cce) [reply crypto/gobatchverifier.go]: Updated

File: crypto/batchverifier_bench_test.go (hunk 8 above)
Comment (cce) [reply crypto/batchverifier_bench_test.go]: this was left over from before when I was just benchmarking one BatchVerifier implementation with no sub-benchmarks.. removed

File: crypto/gobatchverifier.go (hunk 11 above)
Comment (cce) [reply crypto/gobatchverifier.go]: Updated — making this change brought BenchmarkCanonicalityCheck from ~7.2ns/op to ~5.2 ns/op

File: crypto/batchverifier.go [hunk 12]
```diff
@@ -185,18 +204,26 @@ func cgoBatchVerificationImpl(messages []byte, msgLengths []uint64, publicKeys [
...
+		(**C.uchar)(unsafe.SliceData(messages2D)),
+		(**C.uchar)(unsafe.SliceData(publicKeys2D)),
+		(**C.uchar)(unsafe.SliceData(signatures2D)),
+		(*C.uchar)(unsafe.SliceData(messages)),
+		(*C.ulonglong)(unsafe.SliceData(msgLengths)),
+		(*C.uchar)(unsafe.SliceData(publicKeys[0][:])),
+		(*C.uchar)(unsafe.SliceData(signatures[0][:])),
 		C.size_t(numberOfSignatures),
-		(*C.int)(&valid[0]))
+		(*C.int)(unsafe.SliceData(valid)))
```
Comment (cce) [inline crypto/batchverifier.go:217]: @algorandskiy since #5700 Go added `unsafe.SliceData` which is the new recommended way to do this — allows empty slices to still work so `&messages[0]` doesn't panic if it's an empty slice, which I ran into with some test vectors that signed empty messages

File: crypto/gobatchverifier.go (hunk 10 above)
Comment (jannotti) [reply crypto/gobatchverifier.go]: I think you converted it to take the slice because you want wanted to check points that were sitting inside a larger array.  You could use Go's newish ability to cast slices to arrays to do that, but keep the old argument type.

This calls `f` with an array that actually came from inside the slice.


```
func f(a [2]int) {
	fmt.Println(a)
}

func main() {
	arr := [4]int{1, 2, 3, 4}

	// Slice of last two elements
	s := arr[2:]

	// Cast to *[2]int, then dereference to [2]int
	f(*(*[2]int)(s))
}
```

Comment (cce) [reply crypto/gobatchverifier.go]: Well when I comment this out, it does get a little bit slower, so I guess it does optimize those away..

Comment (jannotti) [reply crypto/gobatchverifier.go]: The array code would presumably also optimize them away.

Comment (cce) [reply crypto/gobatchverifier.go]: oh casting slices to arrays does look fun.. OK

Comment (cce) [reply crypto/gobatchverifier.go]: Updated in f317118df but without using `(*(*[32]byte)(x))` instead just doing `([32]byte(x))` or `([32]byte(sig[:32]))`

**Review by jannotti — DISMISSED** [review]

**Review by jannotti — APPROVED** [review]
